### ISSUE TITLE
oCap MVP: Login Per Site

### ISSUE THREAD

**Issue by omnat** [MetaMask/Design#58]
Brief: https://docs.google.com/document/d/1bWfX-3aNezvvQZUaGYhDWWdPs0XPImbttRJte-JCscU/edit

Dev issue: https://github.com/MetaMask/metamask-extension/issues/3383

**Comment by omnat** [comment]
### Strong user insight:
I use different accounts (“identities”) for different dapps. Currently, I name these accounts, so I can switch and login/connect with the correct account.
But, it’d be easier if I didn’t have to remember which account I used for which dapp.

**Use-case: Let user connect Account 1 to Dapp A, Account 3 to Dapp B. Also, don't show my Account 1 address to Dapp B.**

### UX impact
For the user its a simple upgrade to “now I can choose to always login on Peepeth with Account 1, while continue using Account 2 on another tab for paying off bounties. All this without having to interact with MetaMask”. UX improvement is fewer interactions with MM (+ lesser cognitive load and friction). 
It is also an additional use case of “when I switch accounts to use dapp B, every other site I’ve connected to that’s open in another tab now sees this new account” which is less tied to a current need but pretty important for the foundation of safe, distinct identities online

**Comment by omnat** [comment]
Design consideration:
Where does this setting gets indicated - in MM or dapp? 
"In MM, as a user I'd like to then see a list of dapps I am connected to with which account"

**Comment by omnat** [comment]
From design sync:
- UI research to see how other products do this. Like how google allows account sign-in, or how some apps indicate you signed in with a different account etc.

To design this for web3 context, and leverage our understanding of how people use their web3 identities (which are essentially connected to their funds). 
- Research insights around how people tend to use accounts as identities 
- Account management overall can be thought as identity management: 
private view [wallet <> user], outward communication [account / identity <> dapp]

To start the foundations in a human-centric way: Consider use-case jam on dapp-MM interactions.. do we stay general purpose (e.g., China WeChat), or do good in 1 forte like identity. (e.g., Dapper focusing doing games-specific case, or Solutions suggesting enterprise grade MM)

**Comment by bdresser** [comment]
Based on our convo in #80, the ideal Ocaps MVP screen here would contain:

* Select an account (log-in per site)
* Dapp suggests a network to join 
* Open-ended permissions list (small at first)...
  * Prove I own this account (signing data usecase)
  * Future ocaps work (spending limit etc, won't be in production to start)

cc @cjeria @omnat @danfinlay holler with questions!

**Comment by omnat** [comment]
In the scenario where user gives permission to always login to dappA with Account1. 
Which of the following are true and which false?

A) Whenever user opens dappA, irrespective of the account pre-selected in extension, they get logged in  to dappA with Account1. i.e., users aren't expected to go to extension and switch to Account1 anymore.

B) If dappA asks to confirm a tx or confirm a contract interaction, this happens with Account1 too.

C) If user wishes, they can still go to extension and from account switcher menu, switch to an account that is not Account 1.

D) If user opens extension while at dappA tab, they see Account1 selected. If they go to another tab, they see another account selected in extension. 
If this is true: which account is selected on extension if the current tab has a dapp or a web2 app that doesn't have any login-per-site permission? 

@danfinlay Erik (what's his tag?)

**Comment by omnat** [comment]
Made a first pass (with a few options). Ignore visuals, this is super low-fi wireframing (with cheap hacks like screenshots & messed up scales). Just to open discussion, feedback.

Link: https://whimsical.com/R3MpU2jifUGNuAPzLueD5F#VsSo8s35VKHwj9C5rvWesp

Please leave comments on wireframes or on issue.
@cjeria @bdresser @danfinlay

**Comment by rekmarks** [comment]
I think A, B, C as above *should* be how it works. In essence, we do away with the notion of a "globally selected account," and instead define the "selected" or "current" account on a per-dapp basis. At eery stage, we just have to make sure to unambiguously communicate to the user which account is currently being used.

For D, perhaps we don't display an account at all if the user is browsing a website they they aren't "connected" to (dapp or otherwise).

These are opinions.

As for the wireframes: 
- Thank you for making them! It helps me visualize and therefore think about what I'm doing.
- For "connect request initiated" (btw, I'm not sure we should call it "connect" in the permissions model), in the MVP, we are planning to do all-or-nothing permissions for the sake of simplicity, with an account selector dropdown.
- I'm confident that a full-screen connect page will be useful. Consider when the user wants to attenuate/modify permissions somehow (by adding "caveats" in ocap terms); this will likely require more real estate than provided by the popup.
- I really like the beginnings of the "permission management" pages.

Miscellaneous:
- Am I using "pages" correctly? Please do nitpick on my language.
- @omnat, try tagging me by typing out `@rekmarks` in your next comment here and we'll see if that works.

**Comment by bdresser** [comment]
nice work @omnat!

* the "connect" screen looks awesome. i'm cool with all-or-nothing permissions for MVP as @rekmarks says above

* In terms of account management across tabs:
  * I agree A, B, C you describe above should be true
  * I think for D we can show the last account the user was looking at. If I opened the extension on my own and switched to account 2, then dappA switched me to account 1, next time I open the extension (assuming I'm not using dappA in the active tab) then I should see account 2. This is similar to our production behavior now.  

* While connected to a dapp, I think we should
  * include a visual indicator to say "the dapp you're looking at is connected to Account 1" (see image 1)
  * allow switching to other accounts *without* changing the currently connected account
  * include a visual indicator to let users know when they're looking at an account that's not connected to the current dapp (see image 2)

* As privacy mode is implemented right now, it's possible for me to grant access to DappA with two different accounts. When I return to Dapp A on either account, I will not be asked to share my address. I think we should preserve this behavior. That means:
  * users can connect Account 1 to DappA, then later connect Account 2 to DappA
  * when using DappA, both Account 1 & 2 would have the icon shown in image 1
  * when the user navigates to DappA, either 
    * switch the user to the account (1 or 2) they most recently used on DappA, or
    * show a smaller "connect" screen that asks the user to choose between Account 1 & 2

This is a lot! And also just my own opinion, open to other ideas. Let's talk it over in design review tomorrow. 

--

**img 1** - indicate currently connected account
<img width="300" alt="Screen Shot 2019-06-05 at 3 25 21 PM" src="https://user-images.githubusercontent.com/1016190/58994717-58769c80-87a6-11e9-8ab7-cd0a0274a999.png">

**img 2** - indicate when viewing something other than currently connected account (obvi needs some visual love)
<img width="300" alt="Screen Shot 2019-06-05 at 3 28 32 PM" src="https://user-images.githubusercontent.com/1016190/58994795-a12e5580-87a6-11e9-8ba3-b2cd426da50a.png">

mocks in figma [here](https://www.figma.com/file/foWeMNl8hQNDnxBjf2fFcbhd/privacy-mode-notice?node-id=0%3A1)

**Comment by omnat** [comment]
Nice mocks @bdresser !!
And thanks for answering the open questions and adding mocks for those use-cases! Makes sense to me. Keen to discuss it and iron it in design review tomorrow.

**Comment by danfinlay** [comment]
I just wanted to include a link to some of my original thinking on this, which I think still applies very much, and reflects my comments in today's meeting. (sorry for quip links, should move them out).

https://consensys.quip.com/qfEVAnKm3poN
https://consensys.quip.com/PGeGAzVLk3Di
https://consensys.quip.com/YCIOA6RbDUep

**Comment by omnat** [comment]
Another open question I remembered only yesterday was:
**Should we add 'signature request' to the list of this permission?** 

From insight:

> When user logs in some dapps (dapp dependent), they are prompted by 2 MM pop-ups quite back to back: Connect and Signature request. Coupling these as a logging-in action would be useful

**Comment by bdresser** [comment]
True, totally forgot about that @omnat 

q for @danfinlay @rekmarks -- is there a consistent signing format we can provide to dapps that satisfies common auth requirements? would it be acceptable to *not* show the blob of data the user is signing, and just allow dapps to include this as part of an 1102 request invisibly in the background? (perhaps with an option to expand / "see more" to see the exact string being signed?)

**Comment by omnat** [comment]
Next steps (cc @cjeria):
- Answering the open questions about this issue (e.g., Sign request + Connect = Login, edge cases bobby mentioned) 
- Ready for high-fidelity prototype for different use-case (including main edge cases mentioned in this thread)

Wireframes here with Option 3 the most desirable: https://whimsical.com/R3MpU2jifUGNuAPzLueD5F#VsSo8s35VKHwj9C5rvWesp

Omna to share insights on this from user tests.

**Comment by bdresser** [comment]
~I can't access those links @danfinlay - could you copy to this issue or screenshot?~ 

Ah, got in after a million hooops -- here's the third document, most relevant to this: https://docs.google.com/document/d/18IisBHPn-Hs80x5BQ0KYcrLk1zUeIl-TfwTu5xNfG_4/edit?usp=sharing

Main point for discussion is the inclusion of a splash screen that asks the user to select an account, which shifts MetaMask to a true account manager and less of a per-wallet view. Would have ripple effects for other holistic design improvements we've thought of. 

Would really like to keep this conversation going - @cjeria @omnat after spending some time with this please lmk if we should get another discussion on the calendar later this week!

**Comment by omnat** [comment]
@bdresser Splash screen you mentioned is like the latest version in Christian's prototype, right? https://www.figma.com/proto/C8eyTO2iUZc4O83Nd5XsM7yv/oCap-Login-Per-Site?node-id=19%3A1570&scaling=min-zoom. 

From observation in user testing, people (even advanced careful users) tend to not really care much about permissions and go to tap Connect. Pausing is mostly on screens where money is involved. So a 1-step approach is good (not adding extra steps unless user wishes to change the account)

In a user interview, 1 person mentioned they'd choose the hardware wallet account that they use for Uniswap mostly as _Always connect with this account_

> But if my hardware wallet isn't plugged in the next time I join.. What happens then?

**What would happen? Thoughts?** Do we just have an edge case design for that case? 

On the network permission.. 
> Hmm.. not sure if I like this.. Especially when I want to be on Ropsten but it automatically switches me to Mainnet. That wouldn't be good. 

Makes me wonder if we should bring flexibility into this with the checkbox options for additional permissions, instead of all or nothing version? As _'nothing option'_ won't let them interact with the app which is user's main interaction flow. 
**WDYT?**

@danfinlay @cjeria @bdresser ?

**Comment by danfinlay** [comment]
> What would happen? Thoughts? Do we just have an edge case design for that case?

You don't need a hardware wallet connected to select it, only to use it. So for example, if a site _required_ a signature challenge, then you would not be able to log into it with a hardware wallet unless it was connected and supported the signature challenge format (which they tend to be slow to do! Apps should probably stop requiring a signature challenge when they don't really need one!)

**Comment by bdresser** [comment]
> Splash screen you mentioned is like the latest version in Christian's prototype, right? 

No, I mean in the case where the user is NOT connected to a dapp and they open MM, there's an open question: do we want to 
(1) keep the account-specific view we have (in which case we have to decide somehow which account to show), or 
(2) do we zoom one layer out to show them all their accounts and ask which one they're trying to view. This would emphasize the idea that there is no universal *active* account, there's just one per site.

I believe Dan made a rough visual for this but I can't find it now :(

@omnat on that bit of feedback about network switching – sounds like we should make it clearer that we're only switching the network for a particular tab/dapp. If a dapp needs the user to be on a certain network in order to use the dapp, I'm not sure this feedback makes sense? Unless I'm missing something. (Also gut check me @rekmarks @danfinlay that we'll be connecting to different networks on different tabs)

Love the one-screen insight 👍

**Comment by danfinlay** [comment]
> 1 person mentioned they'd choose the hardware wallet account that they use for Uniswap mostly as Always connect with this account

When first signing in, there's a question: Do you have one of their accounts selected like a "default"? Maybe the last one they used? The current prototype we have working just uses the "selected account" as the default one, but has a dropdown for selecting other ones.

> I believe Dan made a rough visual for this but I can't find it now :(

I believe the very rough visual, from that third doc is here. We could probably get by with as little as a "Login" and an "Options" button (both could open full screen mode to different screens):

<img width="250" alt="Screen Shot 2019-06-13 at 5 45 38 PM" src="https://user-images.githubusercontent.com/542863/59476028-9ce7e500-8e03-11e9-9157-8ddfe8696ada.png">

**Comment by cjeria** [comment]
Notes from today's design sync meeting:

### 1. Connect per account, not per wallet
Today, when you connect to a dapp for the first time, you are asked once (per wallet) to connect (connect request pop up). If you switch accounts on the same dapp, you don’t get prompted with a connect request again. 

Screen recording, already connected to dapp, switched accounts no connect request screen.
![](http://g.recordit.co/rPe20lfqyl.gif)

With oCaps log in per site feature, permissions will be set **per account** - since you will eventually be able to give dapps permission to spend up to a certain of value, this permission will only apply for one account and not all accounts in your wallet.

---

### 2. Primary account feature
If more than one account was connected to dapp X, always auto connect with the primary account set by the user. 

Provide a way to set an account as "primary" per dapp. Give users the flexibility to make any account primary

![image](https://user-images.githubusercontent.com/597770/59888025-df20a180-9382-11e9-8dce-418e2698b01d.png)

### 3. Cannot connect two accounts to one dapp at the same time
For this MVP version, we are not designing for the ability to connect two accounts to one dapp at the same time. If the user wants to switch from connected Account 1 to Account 2 (done by switching accounts in the extension) and if it's the first time connecting to account 2, prompt the user with a connect request.

### 4. Disclose permissions directly on Connect screen (not tucked under a link)
We should disclose permission directly on the Connect screen for complete transparency especially when value is associated with permissions. 

### 5. Open question remains: Should we add 'signature request' to the list of this permission? 
https://github.com/MetaMask/Design/issues/58#issuecomment-499629011

#### Link to Prototype
Screen to be updated...
https://www.figma.com/proto/C8eyTO2iUZc4O83Nd5XsM7yv/oCap-Login-Per-Site?node-id=130%3A1097&viewport=-33%2C60%2C0.33460691571235657&scaling=min-zoom

**Comment by danfinlay** [comment]
> Should we add 'signature request' to the list of this permission?

We can add that at any time later, but it mostly needs some cryptographic protocol work, so it should be considered independent and non-blocking of this design issue.

**Comment by danfinlay** [comment]
> 2. Primary account feature

I think we can leave this out of the MVP, it's not necessary at all.

> We should disclose permission directly on the Connect screen for complete transparency especially when value is associated with permissions.

I would argue all permissions are valuable, or they wouldn't be worth getting user consent about. Basically by definition, anything we're requiring consent for is worth showing to the user, not tucking away.

Here's the current state of the working prototype:

![current state of prototype](https://user-images.githubusercontent.com/542863/59889346-d29b4980-9380-11e9-9834-e8cdf792bc3d.gif)

## Other important questions

When you are NOT signed into a site, and you click MetaMask, what do we show? _(I believe this should be a sign-in screen/account selector. Any account details can be tucked behind a "manage accounts" button that opens the full screen interface)_

How does the role of the account menu change when you're signed into a site? Does it now represent changing the account you're signed in with? Does it merely mean looking around at accounts? I think **we need to separate these two things, firmly: Switching accounts should be hidden behind log out, and viewing accounts should be hidden behind another button, maybe "manage accounts", which links to the full screen interface**.

**Comment by bdresser** [comment]
chatting with @danfinlay about the signing challenge as part of this - sounds like this will require some standards work (starting with ETH magicians) and some effort to push that standard through, so we should pursue the MVP as account+network selection and add the signing challenge as a separate feature shortly after.

**Comment by omnat** [comment]
Agree with general direction of tackling signing with standards. 

I'd strongly suggest to combine multiple transactions in a better design (inspiration can be taken from how Google bundles multiple permission asks nicely https://consensys.slack.com/archives/G4V2HTG0Y/p1561659076011900?thread_ts=1561658938.011800&cid=G4V2HTG0Y). This doesn't need any adoption from dapp builders, just an interaction+aesthetic change from our side. cc @cjeria 

This could be a small win for 2 very common cases where 2 or more requests are asked from the user back to back:

- Sign + Connect
- Approve + Confirm (discovered as a pain from Jenny's dev feedbacks as well)

There might be more cases like above..

Maybe this suggestion could be its own separate issue, wdyt @bdresser ?

**Comment by bdresser** [comment]
Yes @omnat this has definitely come up before! See https://github.com/MetaMask/metamask-extension/issues/4050

There's a couple different cases:
* combining non-tx requests (Sign + Connect) which I think is easy from a UX perspective
* combining common pairs of tx (like Approve + Confirm as you mention) which I think is somewhat complex & hard to explain from a UX perspective
* combining multiple arbitrary transactions (like Confirm All for any list of ETH sends or contract calls) which would be quite hard to explain well & design safely

I think prioritizing ocaps work will help solve the original pain point for users. For example, letting users establish a spending limit with dapps would eliminate the need for dapps to request Approve + Confirm transactions in total. So I think the right strategy here is to move forward with a new set of primitives, rather than trying to re-package the building blocks that already exist. 

This might still be worth doing in the meantime, but feels like an interim solution, and I worry it'd take resources away from the larger long-term projects. What do you think?

**Comment by cjeria** [comment]
Posting the latest full tab connect screen designs for discussion:

![image](https://user-images.githubusercontent.com/597770/60827871-c5e76580-a176-11e9-9d92-4fc12c79cc85.png)

**Comment by danfinlay** [comment]
Looking good. A few thoughts:

- We can easily make these line items deselectable, so I think we should reflect that in the design.
- We might want to add the option of embellishing these items with little icons to represent them.
- I feel partial towards selecting the account inline.
- I think we should consider _not_ pre-selecting an account, and making the user deliberately choose one.

## On selecting the account inline

That last screenshot shows the account icon in one place, then the account name and `Switch Accounts` in a different place. I think we could spare some user searching by putting these in one place, maybe allowing the main account item to be selected inline.

It seems a little bit like the current design is prioritizing the current appearance, and this is nice, but given the importance of this screen, I think we should consider what experience we could provide if we departed from it a bit.

Since the account is a very primary thing to select when connecting, maybe it even gets its own screen!

![Screen Shot 2019-07-08 at 11 03 50 AM](https://user-images.githubusercontent.com/542863/60832179-1ad3ad80-a170-11e9-9066-57c2a4c2fa66.png)

A permissions view could appear *after* selecting the account. "New account" could even be an option right there on this screen! Maybe we even have a way to peek at the accounts, what dapps they're connected to, and what assets they're holding.

### Why isolate account selection?

To encourage users to make deliberate choices! If we pre-select an account, we nudge the user towards potentially bad privacy, by just clicking "Confirm" instead of making an important decision! By simply forcing the user to click an account, we encourage a pattern of _deliberate_ disclosure.

I'm not saying it's a _must_, but there are some big benefits to it, and I think we should consider them!

**Comment by wbt** [comment]
[Issue 6903](https://github.com/MetaMask/metamask-extension/issues/6903), in which MetaMask does not allow you to change the account on the transaction confirmation screen after triggering a series of transactions when the wrong account had been selected (but this fact is not visible unless clicking to open the popup before *every* transaction), is related.

**Comment by omnat** [comment]
Updated designs here: https://www.figma.com/proto/C8eyTO2iUZc4O83Nd5XsM7yv/oCap-Login-Per-Site?node-id=257%3A1101&viewport=197%2C572%2C0.30825600028038025&scaling=min-zoom 
cc @rachelcope 

Usability test for these here: https://app.usabilityhub.com/preview/bbdc33ee7973
Doing more research for this with existing users here: https://app.usabilityhub.com/preview/7dd76818382a

Will drop testing links on Monday/Tuesday.

**Comment by wbt** [comment]
Some brief unsolicited feedback:  
![image](https://user-images.githubusercontent.com/563406/62473146-2b2a7880-b76e-11e9-8ddb-09f6f2638fe8.png)  
 - What does "stay connected with Account 2" mean?
 - Does the third permission mean that the dapp will automatically change my MetaMask network selector (i.e. [#5101](https://github.com/MetaMask/metamask-extension/issues/5101))?  Under what conditions will it make that change? Will the connected network be [different for different sites at the same time](https://github.com/MetaMask/metamask-extension/issues/3663)?  Will I have to lose trust in the UI because that behavior changes over time?  
 - What is the relationship between "connect with" and "give permissions to?"
 - As a dapp user, can I choose permissions?  For example, can I not agree to have a dapp change my network selector automatically at unpredictable future times, but still allow the site to view my account's public address and prompt me to sign transactions? 
 - If I only have one account, do I still have to do the inline selection step?

**Comment by danfinlay** [comment]
> What does "stay connected with Account 2" mean?

We can probably remove that line, because it's redundant with "View the public address of Account 2". Please note this is a design, and does not reflect a final list of possible permissions.

> Does the third permission mean that the dapp will automatically change my MetaMask network selector (i.e. #5101)? Under what conditions will it make that change? Will the connected network be different for different sites at the same time? Will I have to lose trust in the UI because that behavior changes over time?

No, we think this would be challenging behavior for users. We are aiming to allow different dapps to connect to different networks simultaneously.

> What is the relationship between "connect with" and "give permissions to?"

They are the same thing. Currently `connect` is the same as `enable()`, and only includes one permission, the permission to view your account. In the future, the same connect screen may include additional permissions.

> As a dapp user, can I choose permissions? For example, can I not agree to have a dapp change my network selector automatically at unpredictable future times, but still allow the site to view my account's public address and prompt me to sign transactions?

Yes, you can uncheck any of the permissions being requested, and we also are interested in allowing users to "attenuate" (adjust terms) of granted permissions.

> If I only have one account, do I still have to do the inline selection step?

You probably will, because it will include an opportunity to create a new account when connecting to a new application. This account selection step may also not be inline, we are considering moving it to an initial screen, since it may have more complexity than the average permission.

**Comment by omnat** [comment]
Per design sync:

- Before sending out the usability test links, we'll update the permissions copy
- Send out the usability test links
- Iterate per findings (if needed)
- Visual design pass in parallel to usability testing (not needed before)

We'll bring this to dev-handoff by end of this sprint.

**Comment by bdresser** [comment]
per today's design sync, this is ready for dev handoff (cc @cjeria @rachelcope)

**Comment by cjeria** [comment]
That's correct! I left notes in the Figma file around the behavior of this feature. Let me know if you have any questions!

https://www.figma.com/file/CvD7nhQVeSEB1QuSy3XhtY/oCap-Testing-Flow?node-id=378%3A4527

![image](https://user-images.githubusercontent.com/597770/64281384-05d18b80-cf21-11e9-9ac2-c4c2730e23a1.png)

**Comment by omnat** [comment]
Update before dev handoff:

- [x] On Connect request screen, update checkboxes to standard ones
- [x] Update this checkboxes in design system too 

@rachelcope

**Comment by cjeria** [comment]
Updated! Link to the latest designs [here](https://www.figma.com/file/C8eyTO2iUZc4O83Nd5XsM7yv/oCap-Login-Per-Site?node-id=614%3A7405)
